### PR TITLE
Chose format from supported ones that we support

### LINKED ISSUE #47585: Vulkan: Editor appears "washed out" or "white-ish" (Mesa drivers)
**Godot version:**
master @ 8f7f5846397297fff6e8a08f89bc60ce658699cc
Built with GCC 10.2

**OS/device including version:**
PopOS 20.04 (ubuntu 20.04)
AMD Radeon RX 5700XT
Intel Haswell CPU

**Issue description:**
The editor appears "white-ish" or "washed out" (see image)
Screenshot_from_2021-04-02_20-18-03

**Steps to reproduce:**
Open editor

**Minimal reproduction project:**
Any project will do, I think, at least all that I've tried.

The problem seems to be here:
https://github.com/godotengine/godot/blob/1328208063de9154019c95b9eae3aaeac50b656b/drivers/vulkan/vulkan_context.cpp#L994

On my system:
```
formatCount == 2
surfFormats[0].format == VK_FORMAT_B8G8R8A8_SRGB
surfFormats[1].format == VK_FORMAT_B8G8R8A8_UNORM
```

Manually setting `format = VK_FORMAT_B8G8R8A8_UNORM` works and the editor is not washed out, but I don't think that's a solution. I noticed this was recently changed in 1328208063de9154019c95b9eae3aaeac50b656b (before that it was `if (true || (formatCount == 1 && surfFormats[0].format == VK_FORMAT_UNDEFINED))` which is debug code, as the commit message implies.)

So the problem appears to be that `VK_FORMAT_B8G8R8A8_SRGB` is not properly handled, at least on my system.

### PR DESCRIPTION
**Pull request by BastiaanOlij**
Check the available formats against formats we support to choose the first match. Basically we want to skip sRGB formats as we're doing the linear -> sRGB color correction ourselves. sRGB formats will do a hardware conversion which doubles this and causes the screen to look washed out.

We may wish to add more supported formats. 

This may fix #47585

### CODE REVIEW

**Comment by BastiaanOlij** [comment]
Note that besides adding more formats to the supported formats list we may wish to order this list from the formats we wish to use and swap the two for loops around. This way we determine priority, not the Vulkan driver which may turn out to just return the list in value order.

File: drivers/vulkan/vulkan_context.cpp [hunk 1]
```diff
@@ -994,11 +994,33 @@ Error VulkanContext::_initialize_queues(VkSurfaceKHR surface) {
...
+
+		// these should be ordered from best to worse,
+		// find the first format that we support,
+		// we want an 32bit RGBA unsigned normalised buffer or similar
+		format = VK_FORMAT_UNDEFINED;
+		for (uint32_t sf = 0; sf < formatCount && format == VK_FORMAT_UNDEFINED; sf++) {
+			for (uint32_t af = 0; af < allowed_formats_count && format == VK_FORMAT_UNDEFINED; af++) {
+				if (surfFormats[sf].format == allowed_formats[af]) {
+					format = surfFormats[sf].format;
+				};
```
Comment (akien-mga) [inline drivers/vulkan/vulkan_context.cpp:1016]:
```suggestion
				}
```

File: drivers/vulkan/vulkan_context.cpp [hunk 2]
```diff
@@ -994,11 +994,33 @@ Error VulkanContext::_initialize_queues(VkSurfaceKHR surface) {
...
+			}
+		}
+
+		if (format == VK_FORMAT_UNDEFINED) {
+			free(surfFormats);
+			ERR_FAIL_V_MSG(ERR_CANT_CREATE, "no usable surface format found");
```
Comment (akien-mga) [inline drivers/vulkan/vulkan_context.cpp:1022]:
```suggestion
			ERR_FAIL_V_MSG(ERR_CANT_CREATE, "No usable surface format found.");
```

**Comment by akien-mga** [comment]
> This may fix #47585

I confirm that this fixes the issue for me with both Intel HD 630 and Radeon RX Vega M running on Mesa 21.1.

**Review by reduz — APPROVED** [review]

**Comment by BastiaanOlij** [comment]
One footnote after discussions with Reduz, we expect most if not all GPUs support `VK_FORMAT_B8G8R8A8_UNORM` so that will likely be selected by default.

If `VK_FORMAT_R8G8B8A8_UNORM` ends up being selected there may be other changes that are required. We'll need to deal with that when someone runs into it.

Same if someone runs into `No usable surface format found.` we'll need to investigate which formats are supported on that device and see if we can add it to the allowed_formats list.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**